File: app/src/main/java/com/shui520it/access/funcion/pool/ShuiThreadPool.java
package com.shui520it.access.funcion.pool;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * @author shuimu{lwp}
 * @time 2019/8/14  11:59
 * @desc
 */
public class ShuiThreadPool {
    private static ShuiThreadPool INSTANCE;
    private ExecutorService executorService;
    public boolean mRunnable = true;

    private ShuiThreadPool() {
        init();
    }

    public static ShuiThreadPool get() {
        if (INSTANCE == null) {
            synchronized (ShuiThreadPool.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ShuiThreadPool();
                }
            }
        }
        return INSTANCE;
    }

    public void init() {
        if (executorService == null) {
            executorService = Executors.newSingleThreadExecutor();
        }
    }

    public synchronized void execute(Runnable run) {
        if (mRunnable) {
            mRunnable = false;
            executorService.execute(run);
        }
    }

    public void shutdown() {
        if (executorService != null) {
            executorService.shutdown();
        }
        executorService = null;
    }

}

File: app/src/main/java/com/shui520it/access/funcion/pool/ShuiThreadPool.kt
package com.shui520it.access.funcion.pool

import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * @author shuimu{lwp}
 * @time 2019/8/14  11:59
 * @desc
 */
object ShuiThreadPool {
    var runnable = true

    /**
     * 一个线程的线程池
     */
    private val executorService: ExecutorService by lazy {
        Executors.newSingleThreadExecutor()
    }

    /**
     * 加上线程安全
     */
    fun execute(run: Runnable) {
        synchronized(ShuiThreadPool) {
            if (runnable) {
                runnable = false
                executorService.execute(run)
            }
        }
    }

    fun shutdown() {
        executorService.shutdown()
    }
}